File: controller/Model/Communication/Request/Sound3dRequest.cs
using System;
using System.Collections.Generic;
using System.Text;
using UltraPlayerController.Communication;
using UltraPlayerController.Model.Enumerations;
using UltraPlayerController.Resources.Communication;
using UltraPlayerController.Model.Utilities;
using Utilities.Serialize;

namespace UltraPlayerController.Model.Communication.Request
{
    public class Sound3dRequest : ARequest
    {
        #region Fields
        private Sound3d _sound3dProperties;
        #endregion

        #region Constructors
        public Sound3dRequest(): base(MessageType.Sound3d)
        {
            _sound3dProperties = new Sound3d();
        }
        #endregion

        #region ARequest members
        public override string ParseToString()
        {
            string result = string.Empty;

            result += base.ParseToString();

            result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, 1, _sound3dProperties.DoplerFactor);
            result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, 1, _sound3dProperties.RolloffFactor);
            result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, 1, _sound3dProperties.MinDistance);
            result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, 1, _sound3dProperties.MaxDistance);
            result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, 1, _sound3dProperties.SourceX);
            result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, 1, _sound3dProperties.SourceY);
            result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, 1, _sound3dProperties.SourceZ);

            return result;
        }
        #endregion

        #region Properties

        public Sound3d Sound3dProperties
        {
            get { return _sound3dProperties; }
            set { _sound3dProperties = value; }
        }
        #endregion

    }
}

File: player/ultraPlayer/Sound3dRequest.cpp
#include "Sound3dRequest.h"

Sound3dRequest::Sound3dRequest(void):ARequest(MessageType::Sound3d)
{
}

Sound3dRequest::~Sound3dRequest(void)
{
}

void Sound3dRequest::FromString( string data )
{
	try
	{
		int paramLength = CommunicationProtocol::Instance()->LongNumericParameterLength();
		int index = CommunicationProtocol::Instance()->HeaderLength();

		_doplerFactor = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_rolloffFactor = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_minDistance = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_maxDistance = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_sourceX = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_sourceY = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_sourceZ = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
	}
	catch (exception& e)
	{
		
	}
}

float Sound3dRequest::DoplerFactor()
{
	return _doplerFactor;
}

float Sound3dRequest::RolloffFactor()
{
	return _rolloffFactor;
}

float Sound3dRequest::MinDistance()
{
	return _minDistance;
}

float Sound3dRequest::MaxDistance()
{
	return _maxDistance;
}

float Sound3dRequest::SourceX()
{
	return _sourceX;
}

float Sound3dRequest::SourceY()
{
	return _sourceY;
}

float Sound3dRequest::SourceZ()
{
	return _sourceZ;
}